roomserver/internal/alias.go

```diff
@@ -91,44 +77,50 @@ func (r *RoomserverAliasAPI) SetRoomAlias(
 	return r.sendUpdatedAliasesEvent(context.TODO(), request.UserID, request.RoomID)
 }
 
-// GetRoomIDForAlias implements alias.RoomserverAliasAPI
-func (r *RoomserverAliasAPI) GetRoomIDForAlias(
+// GetRoomIDForAlias implements alias.RoomserverInternalAPI
+func (r *RoomserverInternalAPI) GetRoomIDForAlias(
 	ctx context.Context,
-	request *roomserverAPI.GetRoomIDForAliasRequest,
-	response *roomserverAPI.GetRoomIDForAliasResponse,
+	request *api.GetRoomIDForAliasRequest,
+	response *api.GetRoomIDForAliasResponse,
 ) error {
 	// Look up the room ID in the database
 	roomID, err := r.DB.GetRoomIDForAlias(ctx, request.Alias)
 	if err != nil {
 		return err
 	}
 
-	if roomID == "" {
-		// No room found locally, try our application services by making a call to
-		// the appservice component
-		aliasReq := appserviceAPI.RoomAliasExistsRequest{Alias: request.Alias}
-		var aliasResp appserviceAPI.RoomAliasExistsResponse
-		if err = r.AppserviceAPI.RoomAliasExists(ctx, &aliasReq, &aliasResp); err != nil {
-			return err
-		}
+	/*
+		TODO: Why is this here? It creates an unnecessary dependency
+		from the roomserver to the appservice component, which should be
+		altogether optional.
```

Comment:
The AS API allows lazy loading of new rooms. See https://matrix.org/docs/spec/application_service/r0.1.2#get-matrix-app-v1-rooms-roomalias

Basically, if the alias doesn't exist, an application service may be able to make it exist so we hit that linked API. Both federated users and local users can hit the alias, so it shouldn't live in `clientapi` or `federationapi`. I think this should probably sit with `publicroomsapi` to be honest, and rename `publicroomsapi` to something like `directoryapi` which:
 - Handles published rooms in the room directory
 - Handles aliases